pkg/core/sync.go

```diff
@@ -0,0 +1,210 @@
+		if err := dropProvider(ctx, storage.DSN(), rp, s.ResourceTables); err != nil {
+			return nil, diag.FromError(fmt.Errorf("drop provider failed: %w", err), diag.INTERNAL)
+		}
+		if err := installProvider(ctx, storage.DSN(), db, want, s.ResourceTables); err != nil {
+			return nil, diag.FromError(fmt.Errorf("install provider failed: %w", err), diag.INTERNAL)
+		}
+	}
+
+	log.Debug().Stringer("provider", opts.Provider).Stringer("state", res.State).Msg("provider sync complete")
+	return res, nil
+}
+
+func Drop(ctx context.Context, storage database.Storage, pm *plugin.Manager, provider registry.Provider) diag.Diagnostics {
+	log.Warn().Stringer("provider", provider).Msg("dropping provider schema")
+	s, diags := GetProviderSchema(ctx, pm, &GetProviderSchemaOptions{Provider: provider})
+	if diags.HasDiags() {
+		return diags
+	}
+	log.Info().Str("provider", provider.Name).Str("version", provider.Version).Msg("dropping provider tables")
+	err := dropProvider(ctx, storage.DSN(), provider, s.ResourceTables)
+	if err != nil {
+		return diag.FromError(err, diag.DATABASE, diag.WithSummary("failed to drop provider"))
+	}
+	return nil
+}
+
+func dropProvider(ctx context.Context, dsn string, provider registry.Provider, resourceTables map[string]*schema.Table) error {
+	dType, _, err := sdkdb.ParseDialectDSN(dsn)
+	if err != nil {
+		return err
+	}
+
+	m, err := migrator.New(logging.NewZHcLog(&log.Logger, "migrator"), dType, nil, dsn, fmt.Sprintf("%s_%s", provider.Source, provider.Name))
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := m.Close(); err != nil {
+			log.Error().Err(err).Msg("failed to close migrator connection")
+		}
+	}()
+
+	if err := m.DropProvider(ctx, resourceTables); err != nil {
+		return err
+	}
+
+	sta, err := state.NewMigratedClient(ctx, dsn, logging.NewZHcLog(&log.Logger, "sync"))
+	if err != nil {
+		return diag.FromError(fmt.Errorf("state failed: %w", err), diag.INTERNAL)
+	}
+	defer sta.Close()
+
+	if err := sta.UninstallProvider(ctx, provider); err != nil {
+		return diag.FromError(fmt.Errorf("state failed: %w", err), diag.INTERNAL)
+	}
+
+	return nil
+}
+
+func installProvider(ctx context.Context, dsn string, db execution.QueryExecer, provider *state.Provider, resourceTables map[string]*schema.Table) error {
+	logger := logging.NewZHcLog(&log.Logger, "sync-install")
+
+	tc := migration.NewTableCreator(logger, schema.PostgresDialect{})
+
+	for _, t := range sort.StringSlice(funk.Keys(resourceTables).([]string)) {
+		up, _, err := tc.CreateTableDefinitions(ctx, resourceTables[t], nil)
+		if err != nil {
+			return diag.FromError(fmt.Errorf("CreateTableDefinitions failed for %s: %w", t, err), diag.INTERNAL)
+		}
+		for _, sql := range up {
+			if err := db.Exec(ctx, sql); err != nil {
+				return diag.FromError(fmt.Errorf("error creating table %s: %w", t, err), diag.INTERNAL)
+			}
+		}
+	}
+
+	sta, err := state.NewMigratedClient(ctx, dsn, logging.NewZHcLog(&log.Logger, "sync"))
```

Comment:
maybe if we touched the state remove the need to pass the logger and use the global one?

Reply:
Removed the logger param from both constructors, now they build one from the global logger in the method. The requirement was mostly because sdkdb and sdkmigrations also require logger in their constructors, which will/might soon go away.